scico/flax/train/apply.py

```diff
@@ -0,0 +1,147 @@
+
+Uses data parallel evaluation.
+"""
+
+from typing import Any, Callable, Optional, Tuple
+
+import jax
+import jax.numpy as jnp
+
+from flax import jax_utils
+
+try:
+    from tensorflow.io import gfile  # noqa: F401
+except ImportError:
+    have_tf = False
+else:
+    have_tf = True
+
+if have_tf:
+    from flax.training import checkpoints
+
+from scico.flax import create_input_iter
+from scico.typing import Array
+
+from .clu_utils import get_parameter_overview
+from .typed_dict import ConfigDict, DataSetDict, ModelVarDict
+
+ModuleDef = Any
+
+
+def apply_fn(model: ModuleDef, variables: ModelVarDict, batch: DataSetDict) -> Array:
+    """Apply current model.
+
+    Assumes sharded batched data and replicated variables for distributed
+    processing.
+
+    This function is intended to be used via
+    :meth:`~scico.flax.only_apply`, not directly.
+
+    Args:
+        model: Flax model to apply.
+        variables: State of model parameters (replicated).
+        batch: Sharded and batched training data.
+
+    Returns:
+        Output computed by given model.
+    """
+    output = model.apply(variables, batch["image"], train=False, mutable=False)
+    return output
+
+
+def only_apply(
+    config: ConfigDict,
+    model: ModuleDef,
+    test_ds: DataSetDict,
+    apply_fn: Callable = apply_fn,
+    variables: Optional[ModelVarDict] = None,
+) -> Tuple[Array, ModelVarDict]:
+    """Execute model application loop.
+
+    Args:
+        config: Hyperparameter configuration.
+        model: Flax model to apply.
+        test_ds: Dictionary of testing data (includes images and labels).
+        apply_fn: A hook for a function that applies current model.
+            Default: :meth:`~scico.flax.train.apply.apply_fn`, i.e. use
+            the standard apply function.
+        variables: Model parameters to use for evaluation. Default:
+            ``None`` (i.e. read from checkpoint).
+
+    Returns:
+        Output of model evaluated at the input provided in `test_ds`.
+
+    Raises:
+        Error if no variables and no checkpoint are specified.
+    """
+    if "workdir" in config:
+        workdir: str = config["workdir"]
+    else:
+        workdir = "./"
+
+    if "checkpointing" in config:
+        checkpointing: bool = config["checkpointing"]
+    else:
+        checkpointing = False
+
+    if variables is None:
+        if checkpointing:
+            if not have_tf:
+                raise RuntimeError(
+                    "Tensorflow not available but is required for Flax checkpointing."
+                )
+            state = checkpoints.restore_checkpoint(workdir, model)
+            variables = {
+                "params": state["params"],
+                "batch_stats": state["batch_stats"],
+            }
+            print(get_parameter_overview(variables["params"]))
+            print(get_parameter_overview(variables["batch_stats"]))
+        else:
+            raise Exception("No variables or checkpoint provided")
```

Comment:
Would a `RunTimeError` not be more appropriate?

Reply:
Updated.